File: hackerearth/goldmines/goldmines.c
#include <stdio.h>
#include <time.h>

int main()
{
  time_t start = time(NULL);
  unsigned int sec;

  // declarations & initialisations

  unsigned short num_rows, num_cols;
  unsigned int num_queries;
  unsigned int num_executions = 0;

  // read the size
  scanf("%hu%hu", &num_rows, &num_cols);

  // build up the accumulated sums
  unsigned long sum_grid[num_rows][num_cols];
  for (unsigned short i=0; i<num_rows; ++i) 
  {
    unsigned long sum = 0;

    for (unsigned short j=0; j<num_cols; ++j) 
    {
      unsigned int gold = 0;
      scanf("%u", &gold);
      sum += gold;

      if (i == 0)
        sum_grid[i][j] = sum;
      else
        sum_grid[i][j] = sum_grid[i-1][j] + sum;
    }
  }

  // read the number of queries
  scanf("%u", &num_queries);
  // prepare the results array
  unsigned long results[num_queries];

  // process the queries
  while (num_executions < num_queries)
  {
    unsigned short x1, y1, x2, y2;
    // read and process the queries
    scanf("%hu%hu%hu%hu", &x1, &y1, &x2, &y2);
      // remember, the grid is zero-indexed
    --x1;
    --y1;
    --x2;
    --y2;

    unsigned long total_gold = sum_grid[x2][y2];

    if (x1 != 0 && x2 != 0) {
      total_gold -= sum_grid[x1-1][y2];
    }
    if (y1 != 0) {
      total_gold -= sum_grid[x2][y1-1];
    }
    if (x1 != 0 && y1 != 0) {
      total_gold += sum_grid[x1-1][y1-1];
    }
    results[num_executions] = total_gold;
    ++num_executions;
  }

  for (unsigned int i = 0; i < num_queries; ++i) {
    fprintf(stdout, "%lu\n", results[i]);
    fflush(stdout);
  }

  sec = (unsigned int) time(NULL) - start;
  fprintf(stdout, "%u seconds\n", sec);

  return 0;
}

// Compiler flags
// gcc -std=gnu99 -w -O2 -fomit-frame-pointer -lm

File: hackerearth/goldmines/goldmines.py
#!/usr/bin/env python

import sys

def run():
  results = []
  input_text = sys.stdin.readlines()

  num_rows, num_cols = map(int, input_text[0].split())
  sum_grid = [[None] * num_cols for r in range(num_rows)]

  for i in range(0, num_rows):
    sum = 0;

    row_gold = map(int, input_text[i+1].split())

    for j in range(0, num_cols):
      sum += row_gold[j];

      if i == 0: #first line of the gold in the grid
        sum_grid[i][j] = sum
      else:
        sum_grid[i][j] = sum_grid[i-1][j] + sum


  num_queries = int(input_text[num_rows + 1])

  for q in range(num_rows+2,num_rows+2+num_queries): #python's range exclude the upper bound
    x1, y1, x2, y2 = map(lambda i: int(i) - 1, input_text[q].split())

    sum = sum_grid[x2][y2]

    if (x1 != 0 and x2 != 0):
      sum -= sum_grid[x1-1][y2]

    if (y1 != 0):
      sum -= sum_grid[x2][y1-1]

    if (x1 != 0 and y1 != 0):
      sum += sum_grid[x1-1][y1-1]

    results.append(sum)

  return results

def print_results(results):
  for r in results:
    print r

results = run()
print_results(results)